main/lsp/LSPPreprocessor.cc

```diff
@@ -86,32 +88,35 @@ void LSPPreprocessor::mergeFileChanges(absl::Mutex &mtx, QueueState &state) {
         const auto &[committed, end] = runningSlowPath.value();
-        // Prune history for all messages no longer in queue or being processed by typechecking thread.
-        ttgs.pruneBefore(min(committed, earliestActiveEditVersion));
-        for (auto it = pendingRequests.begin(); it != pendingRequests.end(); it++) {
-            const auto &msg = *it;
-            if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
-                Timer timeit(logger, "tryCancelSlowPath");
-                auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
-                auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
-                if (combinedUpdates.canTakeFastPath && gs.tryCancelSlowPath(params->updates.versionEnd)) {
-                    logger->debug("[Preprocessor] Canceling typechecking, as edits {} thru {} can take fast path.",
-                                  params->updates.versionStart, params->updates.versionEnd);
-                    params->updates = move(combinedUpdates);
+        earliestActiveEditVersion = min(committed, earliestActiveEditVersion);
+
+        // Avoid canceling if the currently-running slow path has already been canceled.
+        if (!gs.wasTypecheckingCanceled()) {
+            for (auto it = pendingRequests.begin(); it != pendingRequests.end(); it++) {
+                const auto &msg = *it;
+                if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
+                    Timer timeit(logger, "tryCancelSlowPath");
+                    auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
+                    auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
+                    if (combinedUpdates.canTakeFastPath && gs.tryCancelSlowPath(params->updates.versionEnd)) {
+                        logger->debug("[Preprocessor] Canceling typechecking, as edits {} thru {} can take fast path.",
+                                      params->updates.versionStart, params->updates.versionEnd);
+                        params->updates = move(combinedUpdates);
+                    }
+                    break;
+                } else if (!msg->isDelayable()) {
+                    // Message is not delayable, and is not an edit. Can't cancel the slow path.
+                    break;
                 }
-                return;
-            } else if (!msg->isDelayable()) {
-                // Message is not delayable, and is not an edit. Can't cancel the slow path.
-                return;
             }
         }
-    } else {
-        // Prune history for all messages no longer in queue.
-        ttgs.pruneBefore(earliestActiveEditVersion);
     }
+
+    // Prune history for all messages no longer in queue and no longer being typechecked.
+    ttgs.pruneBefore(earliestActiveEditVersion);
```

Comment:
Small cleanup: I changed the control flow slightly to make it clear that pruning happens on all invocations of this method.